src/api/udf/Histogram.h

```diff
@@ -0,0 +1,149 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <atomic>
+#include <glog/logging.h>
+#include <folly/stats/Histogram.h>
+
+#include "surface/eval/UDF.h"
+
+namespace nebula {
+namespace api {
+namespace udf {
+
+template <nebula::type::Kind IK,
+          typename Traits = nebula::surface::eval::UdfTraits<nebula::surface::eval::UDFType::HIST, IK>,
+          typename BaseType = nebula::surface::eval::UDAF<Traits::Type, IK>>
+class Hist : public BaseType {
+  // Set default number of buckets to 20
+  static constexpr size_t BUCKET_NUM = 20;
+
+public:
+  using InputType = typename BaseType::InputType;
+  using NativeType = typename BaseType::NativeType;
+  using BaseAggregator = typename BaseType::BaseAggregator;
+
+public:
+  class Aggregator : public BaseAggregator {
+  static constexpr auto SIZE_SIZE = sizeof(size_t);
+
+  public:
+    explicit Aggregator(InputType min, InputType max, size_t bucketNum = BUCKET_NUM)
+      : min_{ min },
+      max_{ max },
+      bucketNum_{ bucketNum },
+      bucketSize_{ static_cast<InputType>((max_ - min_) / bucketNum) },
+      histogram_{ bucketSize_, min_, max_ } {
+    }
+    virtual ~Aggregator() = default;
+
+    inline virtual void merge(InputType v) override {
+      histogram_.addValue(v);
+    }
+
+    inline virtual void mix(const nebula::surface::eval::Sketch& another) override {
+      auto right = static_cast<const Aggregator&>(another);
+      histogram_.merge(right.histogram_);
+    }
+
+    inline virtual NativeType finalize() override {
+      std::ostringstream tsv;
+      histogram_.toTSV(tsv);
+      return static_cast<NativeType>(tsv.str());
+    }
+
+    // serialize into a buffer
+    inline virtual size_t serialize(nebula::common::ExtendableSlice& slice, size_t offset) override {
+      std::ostringstream tsv;
+      histogram_.toTSV(tsv);
+      auto tsvStr = tsv.str();
+      auto bin = slice.write(offset, tsvStr.size());
+      auto size = slice.write(offset + bin, tsvStr.c_str(), tsvStr.size());
+      return bin + size;
+    }
+
+    inline virtual size_t load(nebula::common::ExtendableSlice& slice, size_t offset) override {
+      auto size = slice.read<size_t>(offset);
+      auto tsvStr = std::string(slice.read(offset + SIZE_SIZE, size));
+      histogram_ = folly::Histogram<InputType>(bucketSize_, min_, max_);
+      // reconstruct histgoram, split by '/n' to get buckets,
+      // then split by '/t' to get min, max, count and sum values
```

Comment:
it would be useful to put a simple example of what the TSV looks like to easier understand below code logic